packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmAny.kt

```diff
@@ -0,0 +1,335 @@
+ * One has to be aware of this when comparing `RealmAny` values generated from different numeral
+ * types, for example:
+ * ```
+ *      RealmAny.create(42.toShort()) == RealmAny.create(42.toByte()) // true
+ * ```
+ * `RealmAny` cannot store `null` values, although `RealmAny` properties **must** be declared
+ * nullable:
+ * ```
+ *      class Warehouse {
+ *          var nonNullableStorage: RealmAny = RealmAny.create("invalid") // This is NOT allowed
+ *          var nullableStorage: RealmAny? = RealmAny.create("valid") // Property MUST be nullable
+ *          var defaultNullStorage: RealmAny? = null // Property MUST be nullable
+ *      }
+ *
+ *      warehouse.nullableStorage = RealmAny.create(22)
+ *      warehouse.nullableStorage = null // Assign null directly to the property
+ * ```
+ * `RealmAny` cannot store [EmbeddedRealmObject]s.
+ *
+ * `RealmAny` values can be sorted. The sorting order used between different `RealmAny` types,
+ * from lowest to highest, is:
+ * - Boolean
+ * - Byte/Short/Integer/Long/Float/Double/Decimal128
+ * - byte[]/String
+ * - Date
+ * - ObjectId
+ * - UUID
+ * - RealmObject
+ *
+ * `RealmAny` properties can be aggregated. [RealmQuery.max] and [RealmQuery.min] produce results
+ * based the sorting criteria mentioned above and thus the output type will be a `RealmAny` instance
+ * containing the corresponding polymorphic value. [RealmQuery.sum] computes the sum of all
+ * numerical values, ignoring other data types, and returns a [Decimal128] result - `SUM`s cannot be
+ * typed-coerced, that is, queries like this are not allowed:
+ * ```
+ *      realm.query<Warehouse>()
+ *          .sum<Float>("nullableStorage") // type CANNOT be coerced to Float
+ * ```
+ */
+public interface RealmAny {
+
+    /**
+     * Supported Realm data types that can be stored in a `RealmAny` instance.
+     */
+    public enum class Type {
+        INT, BOOLEAN, STRING, BYTE_ARRAY, REALM_INSTANT, FLOAT, DOUBLE, OBJECT_ID, REALM_UUID, REALM_OBJECT
+    }
+
+    /**
+     * Returns the [Type] of the `RealmAny` instance.
+     */
+    // TODO use RealmStorageType? How do we avoid problems with RealmStorageType.REALM_ANY?
+    public val type: Type
+
+    /**
+     * Returns the value stored by this `RealmAny` as a [Short]. `RealmAny` instances created using
+     * [Short], [Int], [Byte], [Char] or [Long] values can be converted to any of these types
+     * safely.
+     * @throws [IllegalStateException] if the stored value cannot be safely converted to `Short`.
+     */
+    public fun asShort(): Short
+
+    /**
+     * Returns the value from this `RealmAny` as an [Int]. `RealmAny` instances created using
+     * [Short], [Int], [Byte], [Char] or [Long] values can be converted to any of these types
+     * safely.
+     * @throws [IllegalStateException] if the stored value cannot be safely converted to `Int`.
+     */
+    public fun asInt(): Int
```

Comment:
What about the overflow when the internal `Long` is converted?

Reply:
I added overflow handling. I will add documentation next.